### PR TITLE
patch: add an overlord patch that normalizes channel in state

### PR DESCRIPTION
**Pull request by mvo5**
This PR adds a patch that will normalize the channel in the state.
Snapd was not normalizing the channel so it is possible that the
state contains channels like `/stable` or even `//edge//` without
perceivable ill effects. However with the upcoming change to
"default-tracks" that can be set store side the semantic of those
become less clear. To avoid any ambiguity we normalize the channels
so "/edge" -> "edge".

This can land only *after* we prevent installing from incorrect channels
(like "snap install --channel=/stable").

### CODE REVIEW

File: overlord/patch/patch6_3.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func normChan(in string) string {
+	return strings.Join(strings.FieldsFunc(in, func(r rune) bool { return r == '/' }), "/")
+}
+
+// patch6_3:
+//  - ensure channel spec is valid
+func patch6_3(st *state.State) error {
```
Comment (zyga) [inline overlord/patch/patch6_3.go]: Nitpick, I haven't been in this part of the code lately but I have slight preference for:
```suggestion
func patch6dot3(st *state.State) error {
```

File: overlord/patch/patch6_3.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func normChan(in string) string {
+	return strings.Join(strings.FieldsFunc(in, func(r rune) bool { return r == '/' }), "/")
```
Comment (zyga) [inline overlord/patch/patch6_3.go]: Two comments: please describe in a comment how channel normalisation works and why we need to perform it. It's much better to understand the *intent* than to derive one from the implementation, especially as that latter one drifts.

Second: is this just `filepath.Clean`? I know it's probably not 100% exactly that but it has a well-known semantics and might be a cleaner way to express the intent.

File: overlord/patch/patch6_3.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	// Migrate snapstate
+	dirty := false
+	for name, raw := range snaps {
+		var snapst map[string]interface{}
+		if err := json.Unmarshal([]byte(*raw), &snapst); err != nil {
+			return err
+		}
+		ch := snapst["channel"].(string)
+		if ch != "" {
+			normed := normChan(ch)
```
Comment (pedronis) [inline overlord/patch/patch6_3.go]: given we are here we might as well go all the way  and already parse the channel and use Channel.Clean to get from stable => latest/stable etc (if parsing doesnt fail after basic cleaning).  That kind of state is already possible today if one does "snap install --channel=latest/stable"

**Comment by pedronis** [comment]
Also to be clear. we do need matching code that avoids this kind of values getting back into the state.

**Comment by pedronis** [comment]
marking as blocked for now as discussed in the standup

File: overlord/patch/patch6_3_test.go [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+	// unchanged
+	c.Check(all["track-with-risk"].Channel, Equals, "latest/stable")
+	c.Check(all["track-with-risk-branch"].Channel, Equals, "latest/stable/branch")
+
+	// check tasks
+	task := st.Task("1")
+	c.Assert(task, NotNil)
+	// this was converted
+	var snapsup snapstate.SnapSetup
+	err = task.Get("snap-setup", &snapsup)
```
Comment (zyga) [inline overlord/patch/patch6_3_test.go]: There's an ineffectual assignment here:

```
/home/travis/gopath/src/github.com/snapcore/snapd/overlord/patch/patch6_3_test.go:282:2: ineffectual assignment to err
```

File: overlord/patch/patch6_3.go (hunk 1 above)
Comment (mvo5) [reply overlord/patch/patch6_3.go]: Yeah, it looks a bit out of place but its consistent with the other patch6_{1,2} funcs.

File: overlord/patch/patch6_3_test.go (hunk 4 above)
Comment (mvo5) [reply overlord/patch/patch6_3_test.go]: Thanks, fixed

**Comment by pedronis** [comment]
Related to this @chipaca @mvo5 

> Also to be clear. we do need matching code that avoids this kind of values getting back into the state.

we probably want to proceed this way, rename SnapState.Channel -> SnapState.TrackingChannel, have a SnapState.SetChannel that does cleaning, normalisation

**Comment by pedronis** [comment]
Also I was slightly confused, we don't want cleaning alone, we want Full() .

**Comment by mvo5** [comment]
@chipaca will pick this up, until then I will close this one - he can reopen under his name :)

**Pull request closed without merging**